jupyter_scheduler/task_runner.py

```diff
@@ -0,0 +1,310 @@
+        else:
+            next_run_time = cache.next_run_time
+
+        if should_compute_run_time:
+            self.cache.update(
+                model.job_definition_id,
+                UpdateJobDefinitionCache(
+                    timezone=timezone, next_run_time=next_run_time, active=active, schedule=schedule
+                ),
+            )
+            self.queue.push(
+                JobDefinitionTask(
+                    job_definition_id=model.job_definition_id, next_run_time=next_run_time
+                )
+            )
+        elif active_changed and not active:
+            self.cache.update(model.job_definition_id, UpdateJobDefinitionCache(active=active))
+        elif schedule_changed or timezone_changed:
+            self.cache.update(
+                model.job_definition_id,
+                UpdateJobDefinitionCache(timezone=timezone, schedule=schedule),
+            )
+
+    def delete_job_definition(self):
+        pass
+
+    def pause_jobs(self, job_definition_id: str):
+        self.cache.update(job_definition_id, UpdateJobDefinitionCache(active=False))
+
+    def resume_jobs(self, job_definition_id: str):
+        definition = self.cache.get(job_definition_id)
+        next_run_time = self.compute_next_run_time(definition.schedule, definition.timezone)
+        self.cache.update(
+            job_definition_id, UpdateJobDefinitionCache(active=True, next_run_time=next_run_time)
+        )
+        self.queue.push(
+            JobDefinitionTask(
+                job_definition_id=definition.job_definition_id, next_run_time=next_run_time
+            )
+        )
+
+    def create_job(self, job_definition_id: str):
+        definition = self.scheduler.get_job_definition(job_definition_id)
+        if definition and definition.active:
+            self.scheduler.create_job(
+                CreateJob(**definition.dict(exclude={"schedule", "timezone"}, exclude_none=True))
+            )
+
+    def compute_time_diff(self, queue_run_time: int, timezone: str):
+        local_time = get_localized_timestamp(timezone) if timezone else get_utc_timestamp()
+        return local_time - queue_run_time
+
+    def process_queue(self):
+        while not self.queue.isempty():
+            task = self.queue.peek()
+            cache = self.cache.get(task.job_definition_id)
+            cache_run_time = cache.next_run_time
+            queue_run_time = task.next_run_time
+
+            if not cache.active or queue_run_time != cache_run_time:
+                self.queue.pop()
+                continue
+
+            time_diff = self.compute_time_diff(queue_run_time, cache.timezone)
+            print(f"time difference is : {(time_diff)/1000}")
+
+            # if run time is in future
+            if time_diff < 0:
+                break
+            # if run time is in the past might be in previous run
+            elif time_diff >= (self.run_interval * 1000):
+                break
```

Comment:
I don't understand this block. Why are we halting execution of jobs when we encounter a job that should be run but is within one task runner interval of the current time? I suspect this is a workaround for some runtime bug rather than intended functionality.

Reply:
My initial thought process was avoiding running tasks that were too much in the past because of any runtime situation. But re-visiting this, I am not sure if this condition will ever occur, may be this will happen more if the the run interval is smaller. Since, we are re-hydrate the cache/queue at the start of the server, any past tasks should not appear.

Reply:
My problem with this logic is that you're not popping the item from the heap. If the heap somehow ends up in an invalid state where the top of the heap should have already been run (and thus popped from the heap), the task runner is completely blocked and no new scheduled jobs are run. With the current logic, this exact same condition will be true for all subsequent iterations.

> But re-visiting this, I am not sure if this condition will ever occur, may be this will happen more if the run interval is smaller (1s - 2s). Since, we are re-hydrate the cache/queue at the start of the server, any past tasks should not appear.

Yeah, the cache gets dropped on server shutdown and `next_run_time` is only calculated relative to the current time. Therefore, this condition shouldn't happen to begin with.

I do want to keep this conversation open though, since I did want to call into question whether we should be running past tasks upon server start. For example, if I had a schedule to run on Wednesdays and shut down my server that week, when I start my server again the following Sunday, should it start running my job from last Wednesday? I think we should have a discussion on this.